src/server/raft_client.rs

```diff
@@ -67,136 +82,265 @@ impl Conn {
+            .buffer(mem::replace(&mut conn.buffer, VecDeque::new()))
+            .connect(
+                Arc::clone(&self.env),
+                &self.cfg,
+                Arc::clone(&self.security_mgr),
+                pool,
+            );
+    }
+}
+
+/// `RaftClient` is used for sending raft messages to other stores.
+pub struct RaftClient<S> {
+    ctx: ClientContext<S>,
+    conns: HashMap<(u64, usize), Conn>,
+    pool: CpuPool,
 }
 
-impl RaftClient {
+impl<S: StoreAddrResolver> RaftClient<S> {
     pub fn new(
         env: Arc<Environment>,
         cfg: Arc<Config>,
         security_mgr: Arc<SecurityManager>,
-    ) -> RaftClient {
+        resolver: Arc<S>,
+    ) -> RaftClient<S> {
         RaftClient {
-            env,
+            ctx: ClientContext {
+                env,
+                cfg,
+                security_mgr,
+                resolver,
+                addrs: HashMap::default(),
+                addr_version: 0,
+            },
             conns: HashMap::default(),
-            addrs: HashMap::default(),
-            cfg,
-            security_mgr,
+            pool: Builder::new()
+                .name_prefix("raftclient")
+                .pool_size(1)
+                .create(),
+        }
+    }
+
+    fn conn_mut(&mut self, region_id: u64, store_id: u64) -> &mut Conn {
+        let index = region_id as usize % self.ctx.cfg.grpc_raft_conn_num;
+        match self.conns.entry((store_id, index)) {
+            Entry::Occupied(e) => e.into_mut(),
+            Entry::Vacant(e) => {
+                let conn = self.ctx.connect(&self.pool, store_id);
+                e.insert(conn)
+            }
         }
     }
 
-    fn get_conn(&mut self, addr: &str, region_id: u64, store_id: u64) -> &mut Conn {
-        let index = region_id as usize % self.cfg.grpc_raft_conn_num;
-        let cfg = &self.cfg;
-        let security_mgr = &self.security_mgr;
-        let env = &self.env;
-        // TODO: avoid to_owned
-        self.conns
-            .entry((addr.to_owned(), index))
-            .or_insert_with(|| Conn::new(Arc::clone(env), addr, cfg, security_mgr, store_id))
+    #[inline]
+    pub fn addr(&mut self, store_id: u64) -> Address {
+        self.ctx.addr(store_id)
     }
 
-    pub fn send(&mut self, store_id: u64, addr: &str, msg: RaftMessage) -> Result<()> {
-        let conn = self.get_conn(addr, msg.region_id, store_id);
-        conn.buffer
-            .as_mut()
-            .unwrap()
-            .push((msg, WriteFlags::default().buffer_hint(true)));
+    // TODO: limit the sending messages.
+    pub fn send(&mut self, store_id: u64, msg: RaftMessage) -> Result<()> {
+        let conn = self.conn_mut(msg.region_id, store_id);
+        let msg = (msg, WriteFlags::default().buffer_hint(true));
+        if conn.buffer.len() < MSG_BUFFER_LIMIT {
+            conn.buffer.push_back(msg);
+        } else {
+            // TODO: record metrics.
+            conn.buffer.pop_front();
```

Comment:
Seems there is only 1 producer and 1 concumer. So a lock-free `FIFO` could be very helpful here?

Reply:
I don't think `VecDeque` itself contains any lock.

Reply:
Sorry what I mean is `VecDeque` needs memory re-alloc, however a FIFO buffer doesn't.

Reply:
`VecDeque` itself is a FIFO buffer.

Reply:
@hicqu A `VecDeque` will only alloc if it does not have sufficient capacity, and when it does it allocates a larger block to prevent frequent allocations. :)